File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp = new ListNode(0);
        temp.next = head;
        ListNode x = temp;
        ListNode y = temp;
        for (int i = 1; i <= n + 1; i++) {
            x = x.next;
        }
        while (x != null) {
            x = x.next;
            y = y.next;
        }
        y.next = y.next.next;
        return temp.next;
    }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode temp;
    temp.next = head;
    struct ListNode *x = &temp;
    struct ListNode *y = &temp;
    for (int i = 1; i <= n + 1; i++) {
        x = x->next;
    }
    while (x != NULL) {
        x = x->next;
        y = y->next;
    }
    struct ListNode *to_remove = y->next;
    y->next = y->next->next;
    free(to_remove);
    return temp.next;
}